streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +40,644 @@
+    private void updateLimitOffsetsForStandbyChangelogs(final Map<TopicPartition, Long> committedOffsets) {
+        for (final ChangelogMetadata metadata : changelogs.values()) {
+            if (metadata.stateManager.taskType() == AbstractTask.TaskType.STANDBY &&
+                metadata.stateManager.changelogAsSource(metadata.changelogPartition) &&
+                committedOffsets.containsKey(metadata.changelogPartition)) {
 
-                restoreConsumer.seek(partition, restorer.checkpoint());
-                logRestoreOffsets(partition,
-                        restorer.checkpoint(),
-                        restoreToOffsets.get(partition));
-                restorer.setStartingOffset(restoreConsumer.position(partition));
+                final Long newLimit = committedOffsets.get(metadata.changelogPartition);
+                final Long previousLimit = metadata.restoreLimitOffset;
 
-                log.debug("Calling restorer for partition {}", partition);
-                restorer.restoreStarted();
-            } else {
-                log.trace("Did not find checkpoint from changelog {} for store {}, rewinding to beginning.", partition, restorer.storeName());
+                if (previousLimit != null && previousLimit > newLimit) {
+                    throw new IllegalStateException("Offset limit should monotonically increase, but was reduced for partition " +
+                        metadata.changelogPartition + ". New limit: " + newLimit + ". Previous limit: " + previousLimit);
+                }
 
-                restoreConsumer.seekToBeginning(Collections.singletonList(partition));
-                needsPositionUpdate.add(restorer);
+                metadata.restoreLimitOffset = newLimit;
             }
         }
+    }
 
-        for (final StateRestorer restorer : needsPositionUpdate) {
-            final TopicPartition partition = restorer.partition();
+    private void initializeChangelogs(final Set<ChangelogMetadata> newPartitionsToRestore) {
+        if (newPartitionsToRestore.isEmpty())
+            return;
 
-            // If checkpoint does not exist it means the task was not shutdown gracefully before;
-            // and in this case if EOS is turned on we should wipe out the state and re-initialize the task
-            final StreamTask task = active.restoringTaskFor(partition);
-            if (task.isEosEnabled()) {
-                log.info("No checkpoint found for task {} state store {} changelog {} with EOS turned on. " +
-                        "Reinitializing the task and restore its state from the beginning.", task.id, restorer.storeName(), partition);
+        // for active changelogs, we need to find their end offset before transit to restoring
+        // if the changelog is on source topic, then its end offset should be the minimum of
+        // its committed offset and its end offset; for standby tasks that use source topics
```

Comment:
Can't we just use the committed offset for both? When would the end offset ever be less than the committed?

Reply:
If the source topic is truncated from the tail after committing the position, then it is possible that commit offset > end offset. Since it is an external topic we cannot assume such scenario would never happen.

Reply:
Oof.